numpy/ma/core.py

```diff
@@ -4984,23 +4877,51 @@ def mean(self, axis=None, dtype=None, out=None):
 
         """
         if self._mask is nomask:
-            result = super(MaskedArray, self).mean(axis=axis, dtype=dtype)
-        else:
-            dsum = self.sum(axis=axis, dtype=dtype)
-            cnt = self.count(axis=axis)
-            if cnt.shape == () and (cnt == 0):
-                result = masked
+            ret = super(MaskedArray, self).mean(axis, dtype, out, **kwargs)
+            zerocounts = self.count(axis, **kwargs) == 0
+            if out is not None:
+                if isinstance(out, MaskedArray):
+                    if any(zerocounts):
+                        out.__setmask__(zerocounts)
+                    else:
+                        out.__setmask__(nomask)
+                return out
+            # mask in case of empty slices
+            if any(zerocounts):
+                ret = ret.view(MaskedArray)
+                ret.__setmask__(zerocounts)
+            return ret
+
+        # code largely copied from _methods.mean
+
+        rcount = self.count(axis=axis, **kwargs)
+
+        # Cast bool, unsigned int, and int to float64 by default
+        if dtype is None and issubclass(self.dtype.type,
+                                        (np.integer, np.bool_)):
+            dtype = np.dtype('f8')
+
+        ret = self.sum(axis=axis, dtype=dtype, **kwargs)
+        if ret is masked:
+            return _return_masked(out)
+        with np.errstate(divide='ignore', invalid='ignore'):
+            if isinstance(ret, np.ndarray):
+                ret = np.true_divide(
+                        ret, rcount, out=ret, casting='unsafe', subok=False)
```

Comment:
Thanks for thinking about it again.

I now think the purpose of `subok=False` is to prevent a call to `__array_prepare__`, as that can sometimes prevent the `out` argument from being written to.

Consider
```
class Test(ndarray):
    def __array_prepare__(self, arr, context):
        return np.array(arr)
```
(That appears to be a legitimate way to use `__array_prepare__` judging from a unit test I see). Then do

```
>>> b = np.arange(3).view(type=Test)
>>> r1 = np.add(1, 2, out=b)
>>> b, r1
(Test([0, 1, 2]), Test([3, 3, 3]))
>>> r2 = np.add(1, 2, out=b, subok=False)
>>> b, r2
(Test([3, 3, 3]), Test([3, 3, 3]))
```
So 'out' only gets written to if `subok=False`.

In summary, I think we need to keep `subok=False`.